core/schema/modulesource.go

```diff
@@ -145,25 +163,130 @@ func (s *moduleSchema) moduleSource(ctx context.Context, query *core.Query, args
 	return src, nil
 }
 
+// shared logic with originToPath, copy-pasted
+// to avoid dependency on Buildkit
+func resolveGitURL(ref string) (*gitutil.GitURL, error) {
+	u, err := gitutil.ParseURL(ref)
+	if err != nil {
+		if err == gitutil.ErrUnknownProtocol {
+			u, err = gitutil.ParseURL("https://" + ref)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse URL: %w", err)
+		}
+	}
+	return u, nil
+}
+
 type parsedRefString struct {
 	modPath    string
 	modVersion string
 	hasVersion bool
-	isGitHub   bool
 	kind       core.ModuleSourceKind
+	repoRoot   *vcs.RepoRoot
 }
 
-func parseRefString(refString string) parsedRefString {
+// interface used for host interaction mocking
+type BuildkitClient interface {
+	StatCallerHostPath(ctx context.Context, path string, followLinks bool) (*types.Stat, error)
+}
+
+// parseRefString parses a ref string into its components
+// New heuristic:
+// - stat folder to see if dir is present
+// - if not, try to isolate root of git repo from the ref
+// - if nothing worked, fallback as local ref, as before
+func parseRefString(ctx context.Context, bk BuildkitClient, refString string) (parsedRefString, error) {
 	var parsed parsedRefString
 	parsed.modPath, parsed.modVersion, parsed.hasVersion = strings.Cut(refString, "@")
-	parsed.isGitHub = strings.HasPrefix(parsed.modPath, "github.com/")
 
-	if !parsed.hasVersion && !parsed.isGitHub {
-		parsed.kind = core.ModuleSourceKindLocal
-		return parsed
+	// We do a stat in case the mod path github.com/username is a local directory
+	stat, err := bk.StatCallerHostPath(ctx, parsed.modPath, false)
+	if err == nil {
+		if !parsed.hasVersion && stat.IsDir() {
+			parsed.kind = core.ModuleSourceKindLocal
+			return parsed, nil
+		}
+	}
+
+	// we try to isolate the root of the git repo
+	repoRoot, err := vcs.RepoRootForImportPath(parsed.modPath, false)
+	if err == nil && repoRoot != nil && repoRoot.VCS != nil && repoRoot.VCS.Name == "Git" {
+		parsed.kind = core.ModuleSourceKindGit
+		parsed.repoRoot = repoRoot
+		return parsed, nil
+	}
+
+	parsed.kind = core.ModuleSourceKindLocal
+	return parsed, err
```

Comment:
Hm, is this an error or not?

I don't think we should have this half-errored state, either this should be a "invalid module" error, or we should silently fallback to a local source.

Reply:
Corrected the logic, with the function now logging when these is an error with vcs.RepoRootForImportPath, to be able to hint the user / debug why tge fallback to the local kind is wrong in this case